src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -0,0 +1,848 @@
+            // of that we can hardcode the '>=' operation here.
+            if (it->last_value().has_value() && *it->last_value() >= value) {
+                break;
+            }
+            index += it->size();
+        }
+        if (it != this->_frames.end()) {
+            auto start = const_iterator(it, this->_frames.end(), 0, index);
+            for (; start != this->end(); ++start) {
+                if (pred(*start, value)) {
+                    return start;
+                }
+            }
+        }
+        return this->end();
+    }
+};
+
+/// Column-store iterator
+///
+/// The iterator materializes segment_meta structs when
+/// it's dereferenced. The implementation can be configured
+/// to touch only specific fields of the struct.
+class segment_meta_materializing_iterator
+  : public boost::iterator_facade<
+      segment_meta_materializing_iterator,
+      const segment_meta,
+      boost::iterators::forward_traversal_tag> {
+public:
+    class impl;
+
+    segment_meta_materializing_iterator() = default;
+    segment_meta_materializing_iterator(
+      const segment_meta_materializing_iterator&)
+      = delete;
+    segment_meta_materializing_iterator&
+    operator=(const segment_meta_materializing_iterator&)
+      = delete;
+    segment_meta_materializing_iterator(segment_meta_materializing_iterator&&)
+      = default;
+    segment_meta_materializing_iterator&
+    operator=(segment_meta_materializing_iterator&&)
+      = default;
+    explicit segment_meta_materializing_iterator(std::unique_ptr<impl>);
+
+    ~segment_meta_materializing_iterator();
+
+private:
+    friend class boost::iterator_core_access;
+    const segment_meta& dereference() const;
+
+    void increment();
+
+    bool equal(const segment_meta_materializing_iterator& other) const;
+
+    std::unique_ptr<impl> _impl;
+};
+
+/// Columnar storage for segment metadata.
+/// The object stores segment_meta values using
+/// a dedicated column for every field in the segment_meta
+/// struct
+class segment_meta_cstore {
+    class impl;
+
+public:
+    using const_iterator = segment_meta_materializing_iterator;
+
+    segment_meta_cstore();
+    ~segment_meta_cstore();
+
+    /// Return iterator
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    /// Return last segment's metadata (or nullopt if empty)
+    std::optional<segment_meta> last_segment() const;
+
+    /// Find element and return its iterator
+    const_iterator find(model::offset) const;
+
+    /// Check if the offset is present
+    bool contains(model::offset) const;
+
+    /// Return true if data structure is empty
+    bool empty() const;
+
+    /// Return size of the collection
+    size_t size() const;
+
+    /// Upper/lower bound search operations
+    const_iterator upper_bound(model::offset) const;
+    const_iterator lower_bound(model::offset) const;
+    const_iterator at(size_t ix) const;
+
+    void insert(const segment_meta&);
```

Comment:
We have insert and prefix truncate, but not modifiers for acting in the middle of the list when we do compaction.  Is the idea that those modifications would be implemented externally by rebuilding the whole thing, or would this class get more methods for handling mid-sequence edits?

The compaction case is perhaps the most awkward, so feels important to prove out early.